doc/compiler/optimizer/IntroLoopOptimizations.md

```diff
@@ -0,0 +1,183 @@
+
+### while-do Loop and do-while Loop
+
+Loops are organized in two different ways: (1) while-do loop (2) do-while loop.
+
+A do-while loop executes the body of the loop before it checks the condition. A while-do loop executes the conditions first and then executes the body. The difference between them is that one iteration of the loop is guaranteed to run in a do-while loop. For a while-do loop, it is possible that no single iteration will get to run if the condition check fails.
+
+These two types of loops are canonicalized differently. A do-while loop is in a better state to be optimized than a while-do loop.
+
+For do-while loops, because the condition is checked at the bottom, it is guaranteed that the loop body gets to be executed once. If a computation from the loop body is pulled out into the preheader block, likely it is not going to be any worse than doing the computation inside the loop. In the worst case, the loop will still get to be executed once. Executing those computations in the preheader is no worse than executing them once inside the loop when the loop runs.
+
+Therefore, for a do-while loop, a preheader block can be simply added ahead of the do-while loop body.
+
+For while-do loops, if a computation is added in the preheader from inside a loop, it could become redundant and make things worse if the loop is not entered at all if the condition check fails on the first iteration.
+
+```
+while (i< 15) {
+   z = x + y + i;
+   ++i;
+}
+// Convert a while-do loop into a do-while loop
+if (i< 15) {
+  t = x + y;
+  do {
+     z = t + i;
+     ++i;
+  }
+  while (i<15)
+}
+```
+If it is a while-do loop, we pull the first “if” test that controls whether we enter the loop outside the loop. If it is true, we fall into the loop preheader block and then execute the body of the loop. We will [clone that condition](https://github.com/eclipse/omr/blob/ef9fafc24998865bc4041ad56cc17b2b4c1f210c/compiler/optimizer/LoopCanonicalizer.cpp#L817-L845) so that we check it at the bottom of the loop as well. This is how a while-do loop is converted into a do-while loop by checking the condition outside the loop.
+
+## 2. Loop Versioner
+
+Loop Versioner optimizes away either exception checks or tests from inside the loop so that they will not be executed every time inside the loop. Instead, they will execute less frequently outside the loop. Compared to Loop Canonicalizer, Loop Versioner does something that will improve performance.
+
+There are several different checks, tests, conditions that could be optimized by Loop Versioner. After a loop is transformed by Loop Versioner, there are two versions of a loop: One is a good (fast or hot) version that has all the checks (that meet the necessary conditions/criteria) removed. Another is a bad (slow or cold) version that has all the checks in it.
+
+### An Example of Versioning with Loop Invariant Bound Check
+```
+for (i=0; i<5; i++) {
+   a[i] = i + a[n]; // a[n] is loop invariant
+   }
```

Comment:
I don't think we can claim this. If `n` is in range `[0,3]` then `a[n]` is _not_ loop invariant because its value will change during previous iterations of the loop. You might want to use something like `c[n]` where `a` does not alias `c`.

Reply:
I will change `a[i]` to `c[i]` 
```
c[i] = i + a[n]; // a[n] is loop invariant
```

Reply:
Addressed in 641a90a